docs/code/component-architecture/README.md

```diff
@@ -1,9 +1,132 @@
+
+1. When a component is consumed by another, it doesn't matter what the structure is — the component directory's default export is automatically picked up, as long as the consuming component imports the component from the index file.
+2. It puts all relevant exports in one place. Consider that a test spec could load both the component and the fixture from a single file.
+3. It provides a means to contextually switch the default export depending on the environment. This comes in handy when working with adapters.
+
+### `styles.module.scss`
+
+The stylesheet that serves the component. We typically work with [CSS Modules](https://github.com/css-modules/css-modules), which is supported by default with Gatsby.
+
+These style files are simply [Sass](https://sass-lang.com/) files that are consumed by the component.
+
+### `test.spec.js`
+
+The main test spec for the component. See [the Testing article](/code/testing/) for more information on our approach to testing.
+
+Typically we will test the rendering of the component, along with any logical pieces that it brings.
+
+## Sharing Components & Subcomponents
+
+It's really easy for a project's files to get out of hand if not organized well. Simply dumping components into some `src/components` directory is not scalable. To keep our projects organized, we follow these few principles:
+
+- Components available at the top-most level in the primary components directory are shared (or available to share) among any other component (e.g. `src/components/button/`).
+- Groups of components that serve a purpose other than being a shared component can live in their own top-level source directory. For example, instead of treating templates (which may be components) as a component, they could live in a `src/templates/` directory.
+- Components that exist only as a subcomponent of one other component should live within that directory (e.g. `src/layout/header/`).
+
+For scenarios that fall outside these, use your best judgment. If you're looking for a good example, refer to the [Gatsby Starter](https://github.com/ample/gatsby-starter-ample).
+
+## Naming Conventions
+
+Naming is hard. How about some rules to help?
+
+### Files
+
+Files and directories should be written in kebab-case, which means all lower case with hyphens separating words.
+
+### Components
+
+As much as possible, components should be named for what they are or what they do. Keep it semantic. They should be named in PascalCase and match, as closely as possible, to the filename. For example, if the directory for the component is `solution-finder`, the component object should be `SolutionFinder`.
+
+### Props
+
+We're a little more particular about props:
+
+- Props use snake_case. While this may seem confusing, it is to more easily maintain parity with the naming conventions of the CMS products we tend to use.
```

Comment:
Just to clarify this is intended to be the prop "key"? So `<MyComponent some_prop="value" />` instead of `<MyComponent someProp="value" />`. I'm guessing this is so you can easily spread data from a CMS `<MyComponent {...someCmsData} />`.?

At first glance, it seems awkward (since it goes against React convention) but I think it makes sense since converting case is a little annoying. It also pairs with the style convention to make CSS Modules happy

Reply:
I can be convinced to take this a different direction. I've been exploring the concept of _transformers_ as a way to adjust data coming from the CMS. But I don't have that fully-baked yet. Maybe I finish that work and then come back and we can change the standard?

You're right about snake case with CSS modules, except I don't think that applies to keys, right? We usually just map values to classes, and I wouldn't see a problem with having keys take one approach and values another. That can be contextual depending on the use case.

I'm on the fence. What do you think?

Reply:
@seancdavis I think the transformer is a good idea especially if you're trying to spread data from the CMS that has snake case with keeping the consistency of the JSX. Not sure how difficult a transformation layer is with Gatsby. I rather transform the casing in some way than changing the way props work.

I feel as if camel case is more common in JSX from what I've seen. I'm concerned that things like VS Code plugins will assume camel case when auto-completing and then we have to change to snake case (This may only be a problem for `className` most of the time)

May be worth polling #developers and see what the consensus is.